### PR TITLE
Add support for filtering stacks by organization, tag

### LINKED ISSUE #2712: Support listing stacks that have a given stack tag
The CLI does not yet expose a way to discover which stacks have a given tag.

This could be as simple as `pulumi stack ls --tag mytag`, or even including tags in the results from `pulumi stack ls --json` so that a `jq` command could be used to filter down to the stacks that include a given tag.

### LINKED ISSUE #2769: Add org filter for stack listing
For users in multiple orgs and with lots of stacks `pulumi stack ls -a` can be quite lengthy as it lists _all_ stacks (as expected). This can make the output a bit difficult to parse and also makes it more complex to pipe output to other tools to use for automation. 

`pulumi` should support an `--org` flag (or similar) to filter results to the specified organization.

### PR DESCRIPTION
**Pull request by chrsmith**
Adds new flags to `pulumi stack ls` to support filtering stacks not only by project name, but also organization and tags as well.

This PR creates a new `backend.ListStacksFilter` which is passed to `backend.ListStacks` to wire the new filters through to backend implementations. Currently the `filestate` backend doesn't honor any sort of filtering, while the `httpstate` backend does.

One nuance was that at the CLI command level we only take a single "tag filter", but at the API level we persist a separate tag name and tag value filter. This looks more inline with other systems that support resource tagging. (Though ours is still simplistic compared to say, AWS' support for tagging, since we only support filtering on stack tags for just this one operation, and specifying at most one filter for project/tag/etc.)

The end result is that you can now run `pulumi stack ls --tag "environment=production" --all` and it will return all stacks with an "environment" tag and value "production" that you have access to. Or `pulumi stack ls --tag "gcp:project"` will return all stacks with a tag named "gcp:project" (regardless of value).

Fixes #2712, fixes #2769

### CODE REVIEW

File: cmd/stack_ls.go [hunk 1]
```diff
@@ -88,12 +115,42 @@ func newStackLsCmd() *cobra.Command {
...
+	cmd.PersistentFlags().StringVarP(
+		&projFilter, "project", "p", "", "Filter returned stacks to those with a specific project name")
+	cmd.PersistentFlags().StringVarP(
+		&tagFilter, "tag", "t", "", "Filter returned stacks to those in a specific tag (tag-name or tag-name:tag-value)")
+
 	return cmd
 }
 
+// parseTagFilter parses a tag filter into its separate name and value parts, assuming they are separated by a colon.
+// This requires that the tag name not contain a colon, but the tag's value can. Returns an error if the tag filter is
```
Comment (ellismg) [inline cmd/stack_ls.go]: Don't a majority of the tags we add automatically today contain colons?  Could we use equal or something as a delimiter instead?

Comment (chrsmith) [reply cmd/stack_ls.go]: The equal sign is a _much_ better idea...

**Comment by chrsmith** [comment]
@ellismg Switched to using `tagname=tagvalue`, which is a lot more intuitive.

File: cmd/stack_ls.go [hunk 2]
```diff
@@ -26,21 +27,47 @@ import (
...
-		Args:  cmdutil.NoArgs,
+		Short: "List stacks",
+		Long: "List stacks\n" +
+			"\n" +
+			"This command lists stacks. By default only stacks with the same project name as the\n" +
+			"current workspace will be returned. By passing --all, all stacks you have access to\n" +
+			"will be listed.\n" +
+			"\n" +
+			"Results may be further filtered by passing additional flags. Tag filters may include\n" +
+			"the tag name as well as the tag value, separated by a colon. For example 'environment=production'\n" +
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply cmd/stack_ls.go]: Done.

File: cmd/stack_ls.go [hunk 3]
```diff
@@ -26,21 +27,47 @@ import (
...
+				return errors.Wrap(err, "parsing tag filter")
+			}
+			filter := backend.ListStacksFilter{
+				Organization: orgFilter,
+				Project:      projFilter,
+				TagName:      tagName,
+				TagValue:     tagValue,
+			}
+
+			// If --all is not specified, default to filtering to just the current workspace.
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply cmd/stack_ls.go]: Done.

File: cmd/stack_ls_test.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+		// Error cases
+		{Filter: "=", WantError: true},
+		{Filter: "=no tag name", WantError: true},
+		{Filter: "no tag value=", WantError: true},
+	}
+
+	for _, test := range tests {
+		name, value, err := parseTagFilter(test.Filter)
+
+		var validationErr string
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply cmd/stack_ls_test.go]: Not in particular. I'll make the change as it makes things a little cleaner.

**Comment by ellismg** [comment]
> I think that for now we should enforce in the CLI (and in our service) that if you are filtering by a tag, you provide both the name and the value, which is slightly different from what we have here today.

Per our conversation in `pulumi-service` I think what you have here is good and worthwhile, so please ignore this feedback.

File: cmd/stack_ls.go [hunk 5]
```diff
@@ -88,12 +115,42 @@ func newStackLsCmd() *cobra.Command {
...
+		&orgFilter, "organization", "o", "", "Filter returned stacks to those in a specific organization")
+	cmd.PersistentFlags().StringVarP(
+		&projFilter, "project", "p", "", "Filter returned stacks to those with a specific project name")
+	cmd.PersistentFlags().StringVarP(
+		&tagFilter, "tag", "t", "", "Filter returned stacks to those in a specific tag (tag-name or tag-name=tag-value)")
+
 	return cmd
 }
 
+// parseTagFilter parses a tag filter into its separate name and value parts, assuming they are separated by a colon.
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply cmd/stack_ls.go]: Done.

File: cmd/stack_ls.go [hunk 6]
```diff
@@ -88,12 +115,42 @@ func newStackLsCmd() *cobra.Command {
...
+// This requires that the tag name not contain a colon, but the tag's value can. Returns an error if the tag filter is
+// malformed, such as ":" or ":tag-value". (i.e. a tag name must be specified if a tag value is set.)
+func parseTagFilter(t string) (string, string, error) {
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply cmd/stack_ls.go]: SOO much cleaner. Thanks.

File: cmd/stack_ls.go [hunk 7]
```diff
@@ -88,12 +115,37 @@ func newStackLsCmd() *cobra.Command {
...
+// filter is malformed, such as "=" or "=tag-value". (i.e. a tag name must be specified if a tag value is set.)
+func parseTagFilter(t string) (string, string, error) {
+	parts := strings.SplitN(t, "=", 2)
+	if parts[0] == "" {
+		return "", "", errors.New("no tag name specified")
+	}
+	if len(parts) == 1 {
+		return parts[0], "", nil
+	}
+	if parts[1] == "" {
```
Comment (ellismg) [inline cmd/stack_ls.go]: I think you now want to allow this?

File: cmd/stack_ls.go [hunk 8]
```diff
@@ -26,21 +27,47 @@ import (
...
 	"github.com/pulumi/pulumi/pkg/backend/state"
-	"github.com/pulumi/pulumi/pkg/tokens"
 	"github.com/pulumi/pulumi/pkg/util/cmdutil"
 	"github.com/pulumi/pulumi/pkg/workspace"
 )
 
 func newStackLsCmd() *cobra.Command {
-	var allStacks bool
 	var jsonOut bool
+
```
Comment (ellismg) [inline cmd/stack_ls.go]: NIT: could you remove the blank here so all the values are one after another?  I think that's more in line with other commands.

File: cmd/stack_ls.go [hunk 9]
```diff
@@ -26,21 +27,47 @@ import (
...
+			"\n" +
+			"Results may be further filtered by passing additional flags. Tag filters may include\n" +
+			"the tag name as well as the tag value, separated by an equals sign. For example 'environment=production'\n" +
+			"or just 'devstack-owner'.",
+		Args: cmdutil.NoArgs,
 		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
-			var packageFilter *tokens.PackageName
-			if !allStacks {
+			// Build up the stack filters. If left empty, will be a no-op.
+			tagName, tagValue, err := parseTagFilter(tagFilter)
```
Comment (ellismg) [inline cmd/stack_ls.go]: Doesn't this always return an error when `tagFilter` is empty (which is the default?).  If so, should we only do this when tagFilter is non empty (or change parseTagFilter to return `"", "", nil` in that case)?

**Comment by chrsmith** [comment]
@ellismg PTAL, sorry for the noise here. I changed the underlying data model (`backend.ListStacksFilter`) to use string pointers, as to clearly differentiate between "no filter provided" and (the pathological) "filter all stacks with an empty string for the {project, org, tag}".

**Pull request merged**